compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -1837,47 +1817,52 @@ static LogicalResult setVMVXRootConfigImpl(func::FuncOp entryPointFn,
   return setRootConfigFn(op);
 }
 
-/// Find the root operation for the dispatch region.
+/// Find the root operation for the dispatch region. The priority is:
+///   1. A Linalg operation that has reduction loops.
+///   2. Any other Linalg op.
+///   3. An operation that implements TilingInterface.
+/// If there are multiple operations meeting the same priority, the one closer
+/// to the end of the function is the root op.
 static FailureOr<Operation *> getRootOperation(
     ArrayRef<Operation *> computeOps) {
-  Operation *rootOperation = nullptr;
-  for (auto op : llvm::reverse(computeOps)) {
-    if (auto linalgOp = dyn_cast<linalg::LinalgOp>(op)) {
-      // Do not not treat linalg ops that are all parallel as root operations in
-      // this sweep.
-      if (linalgOp.getNumLoops() == linalgOp.getNumParallelLoops()) continue;
-
-      // All other linalg ops are root ops.
-      rootOperation = op;
-      break;
-    }
+  // TODO(hanchung): The pack ops and unpack ops are unconditionally treated as
+  // root ops. This should be fixed if there are fusion cases.
+  for (auto op : computeOps) {
+    if (isa<tensor::PackOp, tensor::UnPackOp>(op)) return op;
+  }
 
-    if (isa<TilingInterface>(op)) {
-      // All other operations that implement this interface are root ops.
-      rootOperation = op;
-      break;
-    }
+  for (auto op : llvm::reverse(computeOps)) {
+    auto linalgOp = dyn_cast<linalg::LinalgOp>(op);
+    if (!linalgOp) continue;
+    if (linalgOp.getNumReductionLoops()) return op;
   }
-  if (rootOperation) return rootOperation;
 
   // If no root operation is found yet. Look for linalg generic ops.
   for (auto op : llvm::reverse(computeOps)) {
-    if (isa<linalg::LinalgOp>(op)) {
-      rootOperation = op;
-      break;
-    }
+    if (isa<linalg::LinalgOp>(op)) return op;
   }
-  return rootOperation;
+
+  for (auto op : llvm::reverse(computeOps)) {
+    if (isa<TilingInterface>(op)) return op;
+  }
+
+  return nullptr;
 }
 
-/// Finds the root operation in the given list of Linalg operations and sets
-/// its configuration. Returns error for multiple root operations.
-static LogicalResult setRootConfig(func::FuncOp entryPointFn,
-                                   ArrayRef<Operation *> computeOps) {
-  FailureOr<Operation *> rootOp = getRootOperation(computeOps);
-  if (failed(rootOp)) {
-    return failure();
+/// Sets the translation information to use for a dispatch region.
+static LogicalResult setTranslationInfoAndRootConfig(
+    func::FuncOp entryPointFn, ArrayRef<Operation *> computeOps) {
+  // First check if the operations have a preset pipeline.
+  for (auto computeOp : computeOps) {
+    // If the config is preset, do not overwrite it.
+    if (IREE::Codegen::CompilationInfoAttr compilationInfo =
+            getCompilationInfo(computeOp)) {
+      return setUserConfig(entryPointFn, computeOp, compilationInfo);
+    }
   }
```

Comment:
Here is the check. We can only preset configurations through `CompilationInfoAttr`. This is the entry_point that checks if it's preset or not. We set the config directly and early-exit if it's preset.